Build a docker image of the CLI on release (#115)

Lawouach opened the pull request:
Note:

* couldn't run the CLI on Alpine for some unknown reason so using the basic image which is a bit fatter
* this will slow down the travis build on release a bit as the openjdk image isn't small
* how can we overwrite the latest tag?

Lawouach commented:
Mmmh, fair point. 

I'll drop my work then :/

Lawouach commented:
Hopefully that will make more sense :)

Ping @ddgenome

cdupuis commented:
Can we make sure that this maven goal only runs on release builds?

Lawouach commented:
Yeah sorry for the license.

For the volume,  is their difference with having the -v when creating the container? 

For the docs, I have started updating the end-user-documentation to that effect, not the readme here.

cdupuis commented:
Wrt to VOLUME it is considered best practice: see https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/

I still wouldn't mind have some updates to the readme here. At least a pointer to the docs.

kipz commented:
Would be nice to have some docs end up on as per docker.com. E.g.

https://hub.docker.com/_/alpine/

Not sure how this works, or how we become "official"

Lawouach commented:
I've updated as per your comment. The VOLUME directive would only be handy if we were copying data stored in the image in each new container, this isn't the case really so it isn't needed as I understand it.

For the doc, I added a not and pushed a PR onto the doc https://github.com/atomist/end-user-documentation/pull/87

ddgenome commented:
Wouldn't we use the VOLUME instruction to mount `~/.atomist` into the container?

cdupuis commented:
@ddgenome, yes, I think that is what we should do. And I think it is good practice to declare that mount point in the Dockerfile so that people see that they should mount that.

cdupuis commented on pom.xml:
```diff
@@ -359,6 +359,78 @@
+			<plugin>
+				<groupId>net.alchim31.maven</groupId>
+				<artifactId>scala-maven-plugin</artifactId>
+				<version>3.2.2</version>
```
Why do we suddenly have the Scala compiler here? The CLI doesn't have a single Scala class. Please remove.

Lawouach replied:
Ah yeah sorry for that one too.

Lawouach commented:
@ddgenome I don't think that does what you describe here. Exposing a VOLUME instruction only tells docker to create a mount point on the host and, if the directory already exists in the image prior to that instructions, to copy its content into the mount point. However the mount point on the host lives under `/var/lib/docker/containers/VOLUMEHASH` (as per `docker inspect CID` https://docs.docker.com/engine/tutorials/dockervolumes/#/locating-a-volume).

That wouldn't mount your home's directory into the container nor would it mount on the host with the user's permissions. Only `-v` does the former as far as I know. https://docs.docker.com/engine/tutorials/dockervolumes/#/mount-a-host-directory-as-a-data-volume

@cdupuis Fair enough but, though the best practice document is about exposing what the container creates, it's not about sync'ing your home directories per-se.

Overall, depending on how users create their containers, data put on the hosts may or may not be deleted when the container goes away and, if someone wants to simply run rug, this will pollute its disk for no reasons. 

```shell
$ cat Dockerfile 
FROM ubuntu
VOLUME $HOME/.atomist
$ docker build -t with_volume .
Sending build context to Docker daemon 2.048 kB
Step 1/2 : FROM ubuntu
 ---> e4415b714b62
Step 2/2 : VOLUME $HOME/.atomist
 ---> Using cache
 ---> 988c782e548f
Successfully built 988c782e548f
10:04:11 sylvain@sylvain-laptop:/tmp/meh$ docker run --name mytest -it with_volume touch /.atomist/test.txt
10:04:34 sylvain@sylvain-laptop:/tmp/meh$ docker ps -a
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                     PORTS               NAMES
a2ce8749a244        with_volume         "touch /.atomist/t..."   8 seconds ago       Exited (0) 7 seconds ago                       mytest
$ docker volume ls
DRIVER              VOLUME NAME
local               a745c68e2d142c5ff17f3b59f432224f0348ca030e165165d9c44ba67c7efc62
$ docker volume inspect a745c68e2d142c5ff17f3b59f432224f0348ca030e165165d9c44ba67c7efc62
[
    {
        "Driver": "local",
        "Labels": null,
        "Mountpoint": "/var/lib/docker/volumes/a745c68e2d142c5ff17f3b59f432224f0348ca030e165165d9c44ba67c7efc62/_data",
        "Name": "a745c68e2d142c5ff17f3b59f432224f0348ca030e165165d9c44ba67c7efc62",
        "Options": {},
        "Scope": "local"
    }
]
$ sudo ls /var/lib/docker/volumes/a745c68e2d142c5ff17f3b59f432224f0348ca030e165165d9c44ba67c7efc62/_data
test.txt
$ docker rm a2ce8749a244
a2ce8749a244
$ sudo ls /var/lib/docker/volumes/a745c68e2d142c5ff17f3b59f432224f0348ca030e165165d9c44ba67c7efc62/_data
test.txt
$ docker volume rm a745c68e2d142c5ff17f3b59f432224f0348ca030e165165d9c44ba67c7efc62
a745c68e2d142c5ff17f3b59f432224f0348ca030e165165d9c44ba67c7efc62
```

It's only when using `--rm` that the volume is automatically deleted:

```
$ docker run --rm --name mytest -it with_volume touch /.atomist/test.txt
$ docker volume ls
DRIVER              VOLUME NAME
```

ddgenome commented:
Unfortunate. The note at the bottom of this section does indicate you cannot mount host directories in any way from instructions in the Dockerfile https://docs.docker.com/engine/tutorials/dockervolumes/#/mount-a-host-directory-as-a-data-volume

If we do have a VOLUME instruction in the Dockerfile, would the contents of the created volume be persisted from container to container, or would the contents from the image overwrite it every time?

Lawouach commented:
The content would not actually be shared with a single VOLUME instruction. It would persist to the host indeed but each new container would get its own space on the host. I think you can only "reuse" if you are explicit about the host's directory being mounted into the container, via `-v` (which was documented in the branch of the end-user-doc I added).

```
$ docker run --name mytest -it with_volume touch /.atomist/test.txt

$ docker volume ls
DRIVER              VOLUME NAME
local               9ac7587b16bd171a3baf8aca68f41606e6dcff9332df6e6dcb5694c9d042e125

$ docker rm mytest
mytest

$ docker volume ls
DRIVER              VOLUME NAME
local               9ac7587b16bd171a3baf8aca68f41606e6dcff9332df6e6dcb5694c9d042e125

$ docker run --name mytest -it with_volume touch /.atomist/test.txt

$ docker volume ls
DRIVER              VOLUME NAME
local               9ac7587b16bd171a3baf8aca68f41606e6dcff9332df6e6dcb5694c9d042e125
local               bece8b23798272264d1c7360f72d3a8bb509c62a60650972d1936129845161a4
```

cdupuis commented:
The way I did it in the past was to use a `VOLUME /root/.atomist` in the Dockerfile and run with:

`docker run -it -v=~/.atomist:/root/.atomist:rw`

That should give you the same `.atomist` content across restarts of the container.

Lawouach commented:
Indeed, using `-v` will do that. That's the only way, but in that case the VOLUME instruction isn't necessary at all and, if you don't provide `-v`, it will still waste the user's host disk by mounting a random directory that the user will likely not even know about (unless the user inspects the existing volumes through `docker volume ls`).

Would the [section I pushed on the doc](https://github.com/atomist/end-user-documentation/pull/87) be sufficient? We could also add comments in the Dockerfile to that effect.

ddgenome commented on src/main/docker/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+    mkdir -p /home/atomist/.atomist/repository && \
+    mkdir -p /home/atomist/project
+ENV PATH="$PATH:/home/atomist/rug-cli-${version}/bin"
+ADD rug-cli-${version}-bin.tar.gz /home/atomist
```
Don't we also need an instruction to unpack the archive?

Lawouach replied:
That's automatically taken care of by the ADD instruction.

ddgenome replied:
Nice!

Lawouach replied:
Yeah. As far as I understand, it's the only difference with COPY.

ddgenome replied:
As I recall, ADD also allows fetching the file from a URL.

Lawouach replied:
It does indeed.

ddgenome commented on src/main/docker/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+RUN chown -R atomist.atomist /home/atomist
+WORKDIR /home/atomist/project
+USER atomist
+ENTRYPOINT ["rug"]
```
Won't this just run `rug`, which without arguments will print some instructions and exit? For an interactive case like this, isn't it better to avoid `ENTRYPOINT`, rather use `CMD` and provide it with something like:

```
CMD [ "rug", "shell" ]
```

ddgenome replied:
Or even just use use the standard shell so people can follow along with all the other docs as normal.

Lawouach replied:
That sounds fair. I'll update this.

Lawouach commented:
Okay, I think all changes are in.

The only remaining question was to ensure the image is built on release only. I think it is but could you confirm it @ddgenome please?

ddgenome commented on src/main/scripts/travis-build.bash:
```diff
@@ -76,6 +76,11 @@ function main() {
             return 1
         fi
+
+        if ! $mvn package docker:build -DskipTests -DpushImageTag; then
```
It will get built and pushed on every master build and "release". If you only want it on release, put a conditional around the docker build.

```sh
if [[ $TRAVIS_TAG =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
    if ! $mvn package docker:build ...
```

Lawouach replied:
Thanks for the heads up!

Lawouach commented:
Alright, the last build should be the one :)

Lawouach commented:
@kipz there is a discussion there too indeed. We didn't understand the impacts of hosting the image over there I think. Probably worth a discussion.

Would it make us more visible?
